File: tools/bc.py
# bc.py -- Simplistic bytecode packer script
# Description: Takes a list of unsigned integers (range 0-255) and
#              prints them to stdout. 
# Author:  Philip Conrad on 4/30/2013 @ 2:35 PM
# License: MIT-License

"""
How it works:
--------------
Provide the unsigned byte values you wish to pack as a list.
Example:
    byteList = [0,1,2,3, 0,0,1,2, 0,0,0,0]
    result = assemble(byteList)
    print "result:'" + result + "'"

Output to a file:
------------------
Just write the assembled string out to file.

Example:
    byteList = [0,1,2,3, 0,0,1,2, 0,0,0,0]
    result = assemble(byteList)
    outfile = open("bytecode.bc", "w")
    outfile.write(result)
    outfile.flush
    outfile.close
"""


import struct, sys
from ast import literal_eval


# example usage:
# $ python bc.py '[0,2,43,0,  3,2,0,0,  0,1,1,0,  0,0,0,0]'
if __name__ == '__main__':
    try:
        bytelist = literal_eval(sys.argv[1])
        result = [struct.pack("<B", b) for b in bytelist]
        for b in result:
            sys.stdout.write(b)
    except IndexError:
        print "USAGE: python bc.py [BYTE_VALUE_LIST]"
        print "\nExample:\n\tpython bc.py '[0,2,43,0,  3,2,0,0,  0,1,1,0,  0,0,0,0]'"

File: tools/bc.erl
%% @author Philip Conrad <philipaconrad@gmail.com>
%% @copyright 2013 Philip Conrad
%% @license MIT License
%% @doc A simple bytecode assembler in erlang.

-module(bc).
-export([assemble/1]).


%% @doc Turns a list of numbers into a binary.
-spec assemble(list()) -> binary().
assemble([]) -> <<>>;
assemble(L) -> list_to_binary(L).


%% @private eval function returning an erlang term.
mini_eval(S, Environ) ->
    {ok, Scanned, _} = erl_scan:string(S),
    {ok, Parsed}     = erl_parse:parse_exprs(Scanned),
    {value, Term, _} = erl_eval:exprs(Parsed,Environ),
    Term.


%% @private main function for escript.
main(Args) ->
    try
        [HeadArg|_] = Args,
        T = mini_eval(HeadArg, []),
        io:fwrite(standard_io, "~s", [assemble(T)])
    catch
        _:_ -> usage()
    end;

main(_) -> usage().


usage() ->
    io:format("Usage: escript bc.erl ERL_LIST_TERM~n~n"
              "Description:~n    Takes a list of integers (range 0-255) and prints them to stdout.~n~n"
              "Example:~n    escript bc.erl \"[0,1,2,3, 0,0,1,2, 0,0,0,0].\"~n"),
    halt(1).